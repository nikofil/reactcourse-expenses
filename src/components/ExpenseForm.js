import React from 'react'
import { SingleDatePicker } from 'react-dates'
import moment from 'moment'

export default class ExpenseForm extends React.Component {
    constructor(props) {
        super(props)

        const expense = {...props.expense} || {}
        if (expense.createdAt !== undefined) {
            expense.createdAt = moment(expense.createdAt)
        }

        this.state = {
            description: '',
            amount: 0,
            note: '',
            createdAt: moment(),
            error: null,
            ...expense,
        }
    }

    onDescChange = (e) => {
        const description = e.target.value
        this.setState(() => ({description}))
    }

    onAmountChange = (e) => {
        const amount = parseInt(e.target.value)
        if (!isNaN(amount)) {
            this.setState(() => ({amount}))
        }
    }

    onNoteChange = (e) => {
        const note = e.target.value
        this.setState(() => ({note}))
    }

    onDateChange = (createdAt) => {
        this.setState(() => ({createdAt}))
    }

    onSubmit = (e) => {
        e.preventDefault()
        const { id, description, amount, note, createdAt } = this.state
        if (!description || !amount) {
            this.setState(() => ({error: 'Please set description and amount'}))
        } else {
            this.setState(() => ({error: null}))
            const obj = { id, description, amount, note, createdAt: createdAt.valueOf() }
            this.props.submit(obj)
        }
    }

    render() {
        return (
            <div>
            <form onSubmit={this.onSubmit}>
            <div>
            <input type="text" placeholder="Description" value={this.state.description} onChange={this.onDescChange} autoFocus />
            </div><div>
            <input type="number" placeholder="Amount" value={this.state.amount} onChange={this.onAmountChange} />
            </div><div>
            <textarea value={this.state.note} onChange={this.onNoteChange} placeholder="Note"></textarea>
            </div><div>
            <SingleDatePicker date={this.state.createdAt} onDateChange={this.onDateChange} focused={this.state.focused} onFocusChange={ ({focused}) => this.setState({focused}) } />
            </div>
            <p>{this.state.error}</p>
            <input type="submit" value="Submit" />
            </form>
            </div>
        )
    }
}