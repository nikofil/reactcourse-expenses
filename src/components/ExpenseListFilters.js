import React from 'react'
import { connect } from 'react-redux'
import { setTextFilter, sortByAmount, sortByDate, setStartDate, setEndDate } from '../actions/filters'
import { DateRangePicker } from 'react-dates'

class ExpenseListFilters extends React.Component {
    state = { rangeFocus: undefined }

    onDatesChange = ({startDate, endDate}) => {
        this.props.dispatch(setStartDate(startDate))
        this.props.dispatch(setEndDate(endDate))
    }

    render() {
        const props = this.props
        return (
            <div className="content-container">
                <div className="input-group">
                    <div className="input-group__item">
                        <input className="text-input" placeholder="Search expenses" type="text" value={props.filters.text} onChange={(e) => props.dispatch(setTextFilter(e.target.value))} />
                    </div>
                    <div className="input-group__item">
                        <select className="select" value={props.filters.sortBy} onChange={(e) => props.dispatch(e.target.value == 'amount' ? sortByAmount() : sortByDate())}>
                            <option value="date">Date</option>
                            <option value="amount">Amount</option>
                        </select>
                    </div>
                    <div className="input-group__item">
                        <DateRangePicker displayFormat="YYYY/MM/DD" startDate={props.filters.startDate} endDate={props.filters.endDate} onDatesChange={this.onDatesChange} focusedInput={this.state.rangeFocus} onFocusChange={(rangeFocus) => this.setState({rangeFocus})} />
                    </div>
                </div>
            </div>
        )
    }
}

export default connect(({filters}) => ({filters}))(ExpenseListFilters)