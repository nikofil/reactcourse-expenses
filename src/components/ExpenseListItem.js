import React from 'react'

export default function({children, amount, createdAt, remove}) {
    return (<div>
    <h3>{children}</h3>
    <p>{amount} - {createdAt}<small><button onClick={remove}>Remove</button></small></p>
    </div>)
}