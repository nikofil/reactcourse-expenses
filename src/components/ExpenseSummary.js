import React from 'react'

export default function({ expenses }) {
    return (
        <div>
            Viewing { expenses.length } expenses totalling { expenses.map((x) => x.amount).reduce((acc, val) => acc + val) }
        </div>
    )
}