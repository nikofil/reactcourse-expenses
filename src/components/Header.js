import React from 'react'
import { NavLink } from 'react-router-dom'
import { connect } from 'react-redux'
import { startLogout } from '../actions/auth'

export const Header = ({ startLogout }) => (
    <header>
        <h1>Expensify app</h1>
        <NavLink to="/dashboard" activeClassName="is-active">Dashboard</NavLink>
        <NavLink to="/create" activeClassName="is-active">Create</NavLink>
        <NavLink to="/help" activeClassName="is-active">Help</NavLink>
        <button onClick={ startLogout }>Log out</button>
    </header>
)

const mapDispatchToProps = {
    startLogout
}

export default connect(undefined, mapDispatchToProps)(Header)