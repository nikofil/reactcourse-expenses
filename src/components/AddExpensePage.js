import React from 'react'
import { connect } from 'react-redux'
import { startAddExpense } from '../actions/expenses'
import ExpenseForm from './ExpenseForm'

export class AddExpensePage extends React.Component {
    submitHandle = (obj) => {
        this.props.addExpense(obj)
        this.props.history.push('/')
    }

    render() {
        return (
            <div>
                <div className="page-header">
                    <div className="content-container">
                        <h1 className="page-header__title">Add Expense</h1>
                    </div>
                </div>
                <div className="content-container">
                    <ExpenseForm submit={this.submitHandle} />
                </div>
            </div>
        )
    }
}

const mapDispatchToProps = (dispatch) => ({
    addExpense: (obj) => dispatch(startAddExpense(obj))
})
export default connect(undefined, mapDispatchToProps)(AddExpensePage)