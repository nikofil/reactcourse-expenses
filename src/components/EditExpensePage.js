import React from 'react'
import { connect } from 'react-redux'
import ExpenseForm from './ExpenseForm'
import { startEditExpense, startRemoveExpense } from '../actions/expenses';

const EditExpensePage = (props) => {
    const submit = (obj) => {
        props.dispatch(startEditExpense(obj.id, obj)).then(() => (
            props.history.push('/')
        ))
    }

    const remove = () => {
        props.dispatch(startRemoveExpense(props.match.params.id)).then(() => (
            props.history.push('/')
        ))
    }

    return (
        <div>Editing: { props.match.params.id }
            <ExpenseForm expense={props.expense} submit={submit} />
            <button onClick={remove}>Remove</button>
        </div>
    )
}

const mapStateToProps = (state, props) => ({
    expense: state.expenses.find((expense) => expense.id == props.match.params.id)
})

export default connect(mapStateToProps)(EditExpensePage)