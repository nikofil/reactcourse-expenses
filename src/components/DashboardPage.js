import React from 'react'
import { connect } from 'react-redux'
import getVisibleExpenses from '../selectors/expenses';
import ExpenseListItem from './ExpenseListItem';
import ExpenseSummary from './ExpenseSummary';
import ExpenseListFilters from './ExpenseListFilters';
import { removeExpense } from '../actions/expenses';

export const DashboardPage = (state) => {
    const children = state.expenses.map((x, key) => {
        return <ExpenseListItem key={key} {...x}>{x.description}</ExpenseListItem>
    })
    return (
        <div>
            <ExpenseSummary expenses={ state.expenses } />
            <ExpenseListFilters />
            <div>{children}</div>
        </div>
    )
}

const mapStateToProps = ({expenses, filters}) => ({
    expenses: getVisibleExpenses(expenses, filters)
})

export default connect(mapStateToProps)(DashboardPage)