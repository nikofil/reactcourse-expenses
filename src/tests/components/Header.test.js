import React from 'react'
import { shallow } from 'enzyme'
import { Header } from '../../components/Header'

test('should render the Header', () => {
    const wrapper = shallow(<Header />)
    expect(wrapper).toMatchSnapshot()
})

test('should log out on button click', () => {
    const logoutSpy = jest.fn()
    const wrapper = shallow(<Header isAuthenticated={true} startLogout={ logoutSpy } />)
    wrapper.find('button').at(0).simulate('click')
    expect(logoutSpy).toHaveBeenCalled()
})