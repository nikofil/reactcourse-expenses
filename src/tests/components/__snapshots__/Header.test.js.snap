// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render the Header 1`] = `
<header
  className="header"
>
  <div
    className="content-container"
  >
    <div
      className="header__content"
    >
      <Link
        className="header__title"
        replace={false}
        to="/dashboard"
      >
        <h1>
          Expensify
        </h1>
      </Link>
      <button
        className="button"
      >
        Log out
      </button>
    </div>
  </div>
</header>
`;
